tools/timelock_proposal/mcm_with_timelock.go

```diff
@@ -0,0 +1,149 @@
+package timelock_proposal
+
+import (
+	"math/big"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/smartcontractkit/ccip-owner-contracts/tools/errors"
+	owner "github.com/smartcontractkit/ccip-owner-contracts/tools/gethwrappers"
+	"github.com/smartcontractkit/ccip-owner-contracts/tools/mcms_proposal"
+)
+
+var ZERO_HASH = common.BytesToHash([]byte{})
+
+type MCMSWithTimelockChainMetadata struct {
+	mcms_proposal.ChainMetadata
+	TimelockAddress common.Address `json:"timelockAddress"`
+}
+
+type TimelockOperation string
+
+const (
+	Schedule TimelockOperation = "schedule"
+	Cancel   TimelockOperation = "cancel"
+	Bypass   TimelockOperation = "bypass"
+)
+
+type MCMSWithTimelockProposal struct {
+	mcms_proposal.Proposal
+
+	Operation TimelockOperation `json:"operation"` // Always 'schedule', 'cancel', or 'bypass'
+
+	// TODO: this should be configurable as a human-readable string
+	// i.e. 1d, 1w, 1m, 1y
+	MinDelay string `json:"minDelay"`
+
+	// Overridden: Map of chain identifier to chain metadata
+	ChainMetadata map[string]MCMSWithTimelockChainMetadata `json:"chainMetadata"`
+
+	// Overridden: Operations to be executed after wrapping in a timelock
+	Transactions []BatchChainOperation `json:"transactions"`
+}
+
+func (m *MCMSWithTimelockProposal) Validate() error {
+	if err := m.Proposal.Validate(); err != nil {
+		return err
+	}
+
+	switch m.Operation {
+	case Schedule, Cancel, Bypass:
+		break
+	default:
+		return &errors.ErrInvalidTimelockOperation{
+			ReceivedTimelockOperation: string(m.Operation),
+		}
+	}
+
+	_, err := time.ParseDuration(m.MinDelay)
+	if err != nil {
+		return &errors.ErrInvalidMinDelay{
+			ReceivedMinDelay: m.MinDelay,
+		}
+	}
+
+	return nil
+}
+
+func (m *MCMSWithTimelockProposal) ToMCMSOnlyProposal() (mcms_proposal.Proposal, error) {
+	mcmOnly := m.Proposal
+
+	// Start predecessor map with all chains pointing to the zero hash
+	predecessorMap := make(map[string]common.Hash)
+	for chain := range m.ChainMetadata {
+		predecessorMap[chain] = ZERO_HASH
+	}
+
+	// Convert chain metadata
+	mcmOnly.ChainMetadata = make(map[string]mcms_proposal.ChainMetadata)
+	for chain, metadata := range m.ChainMetadata {
+		mcmOnly.ChainMetadata[chain] = mcms_proposal.ChainMetadata{
+			NonceOffset: metadata.NonceOffset,
+			MCMAddress:  metadata.MCMAddress,
+		}
+	}
+
+	// Convert transactions into timelock wrapped transactions
+	for _, t := range m.Transactions {
```

Comment:
We can make this a separate function for better unit testing

Reply:
yeah need to add more extensive unit tests for this regardless but yeah sure